### PR TITLE
(doc): add pull request template

### PR DESCRIPTION
**Pull request by mikemurray**
Adds a pull request template and details to help improve the quality of PR's into core Reaction.

### CODE REVIEW

**Comment by mikemurray** [comment]
@spencern @machikoyasuda your thoughts and reviews are welcome

**Comment by machikoyasuda** [comment]
@zenweasel - What do you think about combining https://github.com/reactioncommerce/reaction-docs/issues/450 issue with this one? This addition adds a checklist for what both the PR author and the Reviewer should do in specific scenarios. You can add anything else in this section: https://github.com/reactioncommerce/reaction-docs/pull/465/files#diff-1b21683f4fb6eabc842c6aa25ccb5411L48


After this PR is merged, we should update the template in https://github.com/reactioncommerce/reaction/blob/master/.github/pull_request_template.md

**Comment by zenweasel** [comment]
@machikoyasuda Yes, #450 should be combined with this one and we should update the template in the main PR repo. Is this the docs that is linked to when you create a PR?

**Comment by mikemurray** [comment]
@zenweasel Yes, the template and a link back to these docs should be in the PR template.

**Comment by machikoyasuda** [comment]
@zenweasel 

The plan is to update https://github.com/reactioncommerce/reaction/blob/master/.github/pull_request_template.md with the content from https://github.com/reactioncommerce/reaction-docs/pull/465/files#diff-1b21683f4fb6eabc842c6aa25ccb5411R34 so that users would see it when they create a PR in `reaction` core.

We should also link this in https://github.com/reactioncommerce/reaction/blob/master/CONTRIBUTING.md from core too, as it is linked in the README: https://github.com/reactioncommerce/reaction#contribute

**Comment by zenweasel** [comment]
@machikoyasuda Just wanted to make sure we were updating this stuff....

**Comment by machikoyasuda** [comment]
@zenweasel That blue `contributing` link === https://github.com/reactioncommerce/reaction/blob/master/CONTRIBUTING.md

**Comment by zenweasel** [comment]
@machikoyasuda got it, thanks

File: developer/contributing.md [hunk 1]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+Resolves #issueNumber
+Impact breaking|minor|bug|style|refactor
+
+## Issue
+
+Description of the issue this PR is solving, why it's happening and how it can be reproduced. This may differ from the original ticket as you now have more information at your disposal.
+
+## Solution
+
+Summarize the your solution to the problem, as well as any things you may have tried along the way that may not have worked.
```
Comment (spencern) [inline developer/contributing.md]: I think this part is really important, but it's a hard to follow (`the your`, `any things`)
What do you think about:
`Summarize your solution to the problem. Please include short descriptions of any solutions you tested before arriving at your final solution. This will help reviewers know why you decided to solve this problem in this particular way and will speed up the review process.`

File: developer/contributing.md [hunk 2]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+
+## Breaking changes
+
+- If you have a breaking changes, list them here, otherwise list none.
```
Comment (spencern) [inline developer/contributing.md]: We may need to give some examples (here or linked) as to what constitutes a breaking change. I'm not sure we have this documented somewhere, but it would be valuable.

File: developer/contributing.md [hunk 3]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+
+## Testing
+
+1. Add steps to test this PR
+1. Be detailed enough that someone can work through it
+1. Avoid being too granular
```
Comment (spencern) [inline developer/contributing.md]: Would be nice to be able to link to a few great example PRs as well, if not from the template itself, then from the contributing guidelines

File: developer/contributing.md [hunk 4]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+##### Top section
+
+###### PR author
+
+The first line of every PR should have `Resolves #ticketNumber`.
+The second line of every PR should have `Impact breaking|minor|bug|style|refactor`.
+
+###### Reviewer
+
+Read over the original ticket, and understand the impact this PR will have on the codebase. If the PR doesn't follow the above template, reject and point the other to this doc.
```
Comment (spencern) [inline developer/contributing.md]: `the above template` => `our template`?

`point the other to this doc` => `point to this doc` or `point the owner to this doc`?

File: developer/contributing.md [hunk 5]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+
+###### Reviewer
+
+Use this information to help determine path to test this PR. Research any included packages or techniques that may have been used that you're not familiar with. Ask questions if you're confused.
+
+##### Breaking changes
+
+###### PR author
+
+List any breaking changes that come with this PR. If you have none, then simply put `none`. Examples of breaking changes include changing file names, moving files around, deleting files, renaming functions or exports, or significantly changing code to where previous versions of the app may not work as expected.
```
Comment (spencern) [inline developer/contributing.md]: This may be a good enough example list for now. Should we include this in the template? Not sure how verbose the template should be. @machikoyasuda any thoughts?

Comment (spencern) [reply developer/contributing.md]: Might be worth instructing the PR author to ensure that she has migrations or deprecation warnings for any breaking changes that were added.

File: developer/contributing.md [hunk 6]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+
+###### Reviewer
+
+If there are breaking changes, make you'll need to make sure any necessary migrations or guards have been put into place so existing apps do not break. Try testing this PR on an existing version of reaction you may already have data for.
```
Comment (spencern) [inline developer/contributing.md]: `Reaction` should always be capitalized as a proper noun.
`Try testing this PR on an existing version of reaction you may already have data for` is a little difficult to follow.

maybe something like
`Test applying this patch to an existing install of Reaction with existing users, orders, carts, etc. Specifically, test any parts of the app where the breaking change is involved and any data set that is involved in a migration.

File: developer/contributing.md [hunk 7]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+
+##### Testing
+
+###### PR author
+
+List the steps needed for testing your change in this section. Assume that testers already know how to start the app, and do the basic setup tasks. Beyond that, the steps should guide users through the feature or fix you've implemented in this PR. Include additional steps if multiple paths are available.
+
+###### Reviewer
+
+Run through the author's steps to verify that it works as they've tested it. Then run through the app on your own as you would test it. Run through the app as many times as you feel comfortable before issuing an approval or a request for changes.
```
Comment (spencern) [inline developer/contributing.md]: `before issuing an approval or a request for changes` => before approving or requesting changes`

File: developer/contributing.md [hunk 8]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
 
+### Pull request template
+
+```
+Resolves #issueNumber
+Impact breaking|minor|bug|style|refactor
+
+## Issue
+
+Description of the issue this PR is solving, why it's happening and how it can be reproduced. This may differ from the original ticket as you now have more information at your disposal.
```
[the remark this answers is not in the document]
Comment (mikemurray) [reply developer/contributing.md]: Isn't this the solution section?

File: developer/contributing.md (hunk 3 above)
Comment (mikemurray) [reply developer/contributing.md]: catch 22, we'd have to use the template in a PR before we could link it here

File: developer/contributing.md (hunk 1 above)
Comment (mikemurray) [reply developer/contributing.md]: 👍

File: developer/contributing.md (hunk 2 above)
Comment (machikoyasuda) [reply developer/contributing.md]: Breaking change example: https://github.com/reactioncommerce/reaction/pull/3331

Comment (machikoyasuda) [reply developer/contributing.md]: https://github.com/reactioncommerce/reaction/pull/3524

File: developer/contributing.md (hunk 3 above)
Comment (machikoyasuda) [reply developer/contributing.md]: Maybe we can combine `Be detailed enough so someone can work through it, while not being too granular.`

File: developer/contributing.md (hunk 5 above)
Comment (zenweasel) [reply developer/contributing.md]: I would specifically call out Schema changes as an example of a breaking change

File: developer/contributing.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (zenweasel) [reply developer/contributing.md]: We should specifically say something like "you must include all information here so that the PR reviewer can review this PR without reading the original issue". So no "see issues for test instructions"

[the remark this answers is not in the document]
Comment (zenweasel) [reply developer/contributing.md]: Not sure where best to put this but something like "Failure to follow this template may result in your PR being closed without comment"

File: developer/contributing.md [hunk 9]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+##### Top section
+
+###### PR author
+
+The first line of every PR should have `Resolves #ticketNumber`.
+The second line of every PR should have `Impact breaking|minor|bug|style|refactor`.
+
+###### Reviewer
+
+Read over the original ticket, and understand the impact this PR will have on the codebase. If the PR doesn't follow the our template, reject and point the author of the PR to this doc.
```
Comment (zenweasel) [inline developer/contributing.md]: I'm confused by this section. Can we just have one PR Author and one PR Review section rather than these alternating ones? Am I the only one bothered by this? One is a list of drop-dead requirements and one is a list of "things to consider", I'm not sure if they are really equivalent?

File: developer/contributing.md [hunk 10]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+##### Issue
+
+###### PR author
+
+Describe the issue this PR is solving with the knowledge you've gained by fixing it. This may differ from the original ticket as you now have more information at your disposal.
+
+###### Reviewer
+
+Use this information as the basis for your review. If something is not clear here, Reject the PR and ask for clarity. While the original issue may have useful information, the PR should be the most up to date representation if ideas.
+
```
Comment (zenweasel) [inline developer/contributing.md]: typo: "of ideas"

File: developer/contributing.md [hunk 11]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+###### Reviewer
+
+Test by applying this patch to an existing install of Reaction with existing users, orders, carts, etc. Specifically, test any parts of the app where the breaking change is involved and any data set that is involved in a migration.
+
+##### Testing
+
+###### PR author
+
+List the steps needed for testing your change in this section. Assume that testers already know how to start the app, and do the basic setup tasks. Beyond that, the steps should guide users through the feature or fix you've implemented in this PR. Include additional steps if multiple paths are available.
```
Comment (zenweasel) [inline developer/contributing.md]: "Also point out areas where your change may have broken existing functionality and list those areas as well. A good test plan created by the Author is one of the keys to maintaining quality and it's in your interest to make your test plan as complete as possible"

Comment (zenweasel) [reply developer/contributing.md]: In super big letters "If your PR fails the basic test instructions your PR may be closed and may receive no further consideration". In other words, if you give me one thing to test on your PR, and even *that* doesn't work, I am not wasting any more time on your PR

File: developer/contributing.md (hunk 9 above)
Comment (zenweasel) [reply developer/contributing.md]: Somewhere (I can't tell where in this current format) we need to have a checklist of things code reviewers should consider.

1. Performance (does the code use best practices for minimizing the load on client and server hardware)
1. Concurrency (are data transactions multi-user safe? Am I prevented from accidentally overwriting someone else changes?)
1. Security (are permissions being checked everywhere appropriate? Are error messages being sanitized before they reach the client. Are we checking inputs properly wherever we accept client input?) 

These are not a comprehensive lists. But we should start compiling them.

File: developer/contributing.md [hunk 12]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+
+###### Reviewer
+
+Use this information as the basis for your review. If something is not clear here, Reject the PR and ask for clarity. While the original issue may have useful information, the PR should be the most up to date representation if ideas.
+
+##### Solution
+
+###### PR author
+
+Summarize your solution to the problem. Please include short descriptions of any solutions you tested before arriving at your final solution. This will help reviewers know why you decided to solve this problem in this particular way and will speed up the review process.
```
Comment (zenweasel) [inline developer/contributing.md]: Maybe just "Summarize how you fixed the problem. Please include short descriptions of other things you tried before arriving at your final solution. "
Just to cut down on the repetition of the word "solution"?

File: developer/contributing.md [hunk 13]
```diff
@@ -31,6 +31,84 @@ Start small and open up a dialogue with us. This will help to get your contribut
...
+
+###### Reviewer
+
+Use this information to help determine path to test this PR. Research any included packages or techniques that may have been used that you're not familiar with. Ask questions if you're confused.
```
Comment (zenweasel) [inline developer/contributing.md]: typo: "determine the path"

File: developer/contributing.md (hunk 12 above)
Comment (zenweasel) [reply developer/contributing.md]: "If you have introduced new dependencies, explain why and how you determined that said dependency was the best solution to the problem."
